asmcomp/debug/available_regs.ml

```diff
@@ -0,0 +1,387 @@
+                let remains_available =
+                  live_across
+                    || (holds_immediate && on_stack)
+                in
+                not remains_available)
+              avail_before
+          | _ -> RD.Set.empty
+        in
+        let made_unavailable =
+          RD.Set.union made_unavailable_1 made_unavailable_2
+        in
+        (* Subtract out the registers made unavailable; then add in the
+           registers that are made *available* -- namely the results of the
+           operation. *)
+        Ok (RD.Set.union
+          (RD.Set.without_debug_info (Reg.set_of_array instr.res))
+          (RD.Set.diff avail_before made_unavailable))
+      | Iifthenelse (_, ifso, ifnot) -> join [ifso; ifnot] ~avail_before
+      | Iswitch (_, cases) -> join (Array.to_list cases) ~avail_before
+      | Iloop body ->
+        let avail_after = ref (RAS.Ok avail_before) in
+        (* XCR pchambart: This should probably be instr.available_before to do a
+           single loop in case of nested Iloop.
+           mshinwell: I _think_ this is fixed, but not completely sure what
+           was being referred to.  Please check. *)
+        begin try
+          while true do
+            (* XCR pchambart: I don't see any situations where a second
+               round would remove more registers from the set. Hence
+               I'm not convinced that there are situations where a
+               fixpoint is required. But this is necessary for marking
+               each instruction in the loop with its real
+               available_before. An assertion should check that never
+               more than 2 rounds are required.
+               mshinwell: I suspect this isn't the case with multiple
+               handlers.  (This Iloop case should go away soon.) *)
+            let avail_after' =
+              RAS.inter !avail_after
+                (available_regs body ~avail_before:!avail_after)
+            in
+            if RAS.equal !avail_after avail_after' then begin
+              raise Exit
+              end;
+            avail_after := avail_after'
+          done
+        with Exit -> ()
+        end;
+        Unreachable
+      | Icatch (recursive, handlers, body) ->
+        List.iter (fun (nfail, _handler) ->
+            (* In case there are nested [Icatch] expressions with the same
+               handler numbers, we rely on the [Hashtbl] shadowing
+               semantics. *)
+            Hashtbl.add avail_at_exit nfail RAS.Unreachable)
+          handlers;
+        let avail_after_body =
+          available_regs body ~avail_before:(RAS.Ok avail_before)
+        in
+        let aux (nfail, handler) (nfail', avail_at_top_of_handler) =
+          assert (nfail = nfail');
+          (* This [available_regs] call may cause [avail_at_exit] to be
+             updated. *)
+          available_regs handler ~avail_before:avail_at_top_of_handler
```

Comment:
To avoid useless traversal (with a potential linear speedup if the handlers are in the wrong order), this could check if the argument changed. i.e. if avail_at_top_of_handler is equal to handler.available_before.

Reply:
This check could be done for every instruction to avoid too much recomputing in case of nested fixpoints. But that would require an efficient equality test.

Reply:
Hmm, except you need to get the "available_after" set somehow.  I think for the moment I'm going to leave this, I have added a comment in the code referencing these suggestions.